test/unit_tests/threads/test_multicore.cpp

```diff
@@ -0,0 +1,248 @@
+#include <catch2/catch.hpp>
+
+#include <hocdec.h>
+#include <ocfunc.h>
+#include <code.h>
+#include <thread>
+#include <algorithm>
+#include <vector>
+#include "../unit_test.h"
+
+#ifdef NRN_ENABLE_THREADS
+#include <multicore.h>
+#include <nrn_ansi.h>
+
+extern int use_cachevec;
+extern int nrn_how_many_processors();
+
+/* @brief
+ *  Test multicore implementation:
+ *  * parallel mode (std::threads)
+ *  * parallel mode with busywait
+ *  * serial mode
+ *  * performance
+ *      * NOTE: GitHub runners don't have enough capabilities for performance KPIs
+ */
+
+
+// generate a range starting from 1 and doubling until we reach the nof concurrent threads
+auto make_available_threads_range() {
+    auto nof_threads_range =
+        std::vector<int>(1 + static_cast<int>(std::log2(nrn_how_many_processors())), 1);
+    std::generate(nof_threads_range.begin() + 1, nof_threads_range.end(), [n = 1]() mutable {
+        return n *= 2;
+    });
+    return nof_threads_range;
+}
+TEST_CASE("Multicore unit and performance testing", "[NEURON][multicore]") {
+    SECTION("Simulation set-up", "[NEURON][multicore][setup]") {
+        GIVEN("We create a parallel context") {
+            REQUIRE(hoc_oc("objref pc\n"
+                           "pc = new ParallelContext()") == 0);
+            THEN("we assert nrn_thread is equal to 1") {
+                REQUIRE(nrn_nthread == 1);
+            }
+            THEN("we check we have no worker threads") {
+                REQUIRE(nof_worker_threads() == 0);
+            }
+            THEN("we setup 1000 passive membrane cells for the simulation") {
+                REQUIRE(hoc_oc(pass_cell_template) == 0);
+                REQUIRE(hoc_oc(prun) == 0);
+                std::string cells{1000_pas_cells};
+                REQUIRE(hoc_oc(cells.c_str()) == 0);
+            }
+        }
+    }
+
+    SECTION("Test parallel mode", "[NEURON][multicore][parallel]") {
+        const auto nof_threads_range = make_available_threads_range();
+        static std::vector<double> cache_sim_times;
+        static std::vector<double> no_cache_sim_times;
+        GIVEN("we do prun() over each nof_threads{nof_threads_range} with cachevec{0,1}") {
+            auto cache_efficient = GENERATE(0, 1);
+            auto nof_threads = GENERATE_REF(from_range(nof_threads_range));
```

Comment:
I **think** `GENERATE_COPY` might be safer here? IIUC this captures a reference to `nof_threads_range` that then goes out of scope, and it just works because the next time this code is executed then a new copy of `nof_threads_range` has (probably) been created in the same place.

IIRC, the pattern here:
https://github.com/neuronsimulator/nrn/blob/140d92f989f3f4cd8cf7c441e225744e9cd747f1/test/unit_tests/container/node.cpp#L55-L57
was an attempt to avoid a similar issue with `GENERATE_REF`.

Perhaps making `nof_threads_range` be `static` would also help?

Reply:
Thanks, will look into it